pkg/terraform/exec/plugins/plugin_versions_generate.go

```diff
@@ -0,0 +1,249 @@
+
+		// If this is a v2 or up  module, chop off the end
+		if pre.Match([]byte(mod.Path)) {
+			path := ure.FindString(mod.Path)
+			url = fmt.Sprintf("https://%s", path)
+		}
+
+		// git clone if directory doesn't exist, else git pull
+		tmpDir := filepath.Join("tmp", p)
+		if _, err := os.Stat(tmpDir); os.IsNotExist(err) {
+			os.MkdirAll(tmpDir, 0700)
+			repo, err = git.PlainClone(tmpDir, false, &git.CloneOptions{
+				URL:      url,
+				Progress: os.Stdout,
+			})
+			if err != nil {
+				log.Fatal("failed to clone repo ", url)
+			}
+		} else {
+			repo, err = git.PlainOpen(tmpDir)
+			if err != nil {
+				log.Fatal("failed to open repo ", tmpDir)
+			}
+			worktree, err := repo.Worktree()
+			if err != nil {
+				log.Fatal(err)
+			}
+			err = worktree.Pull(&git.PullOptions{})
+			if err != nil && err != git.NoErrAlreadyUpToDate {
+				log.Fatal(err)
+			}
+		}
+
+		pi.Tag = mod.Version
+
+		// If the version is a pseudo version, get the commit hash and
+		// resolve it to a tag if one exists, else default to v1.0.0
+		if tre.Match([]byte(mod.Version)) {
+			h := hre.FindString(mod.Version)
+			hash, err := repo.ResolveRevision(plumbing.Revision(h))
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			here, err := os.Getwd()
+			if err != nil {
+				log.Fatal(err)
+			}
+			err = os.Chdir(tmpDir)
+			if err != nil {
+				log.Fatal(err)
+			}
+			// XXX: HACK: fork git process until go-git supports name-rev
+			out, err := exec.Command("git", "name-rev", "--tags", "--name-only", hash.String()).Output()
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			// XXX:  What should be the default here?
+			tag := ore.ReplaceAllString(string(out), "")
+			if tag == "undefined" {
+				tag = "v1.0.0"
+			}
+			pi.Tag = tag
+
+			err = os.Chdir(here)
+			if err != nil {
+				log.Fatal(err)
+			}
+		}
+
+		// Extract version out of tag
+		pi.Version = vre.FindString(pi.Tag)
+	}
+}
+
+// getLocalTerraformPluginVersions gathers the terraform provider version
+// numbers for local plugins
+func getLocalTerraformPluginVersions(file *modfile.File, plugins *map[string]*TFPluginInfo) {
+	(*plugins)["azureprivatedns"] = &TFPluginInfo{
+		Name:    "azureprivatedns",
+		Tag:     "v1.0.0",
+		Version: "1.0.0",
+	}
+	(*plugins)["vsphereprivateprovider"] = &TFPluginInfo{
+		Name:    "vsphereprivateprovider",
+		Tag:     "v1.0.0",
+		Version: "1.0.0",
+	}
+}
+
+func main() {
+	if len(os.Args) != 3 {
+		log.Fatal("usage: ", filepath.Base(os.Args[0]), " <go.mod> <output>")
+	}
+
+	// Get absolute paths for file arguments
+	goMod, err := filepath.Abs(os.Args[1])
+	if err != nil {
+		log.Fatal("failed to get absolute path of go.mod")
+	}
+	outFile, err := filepath.Abs(os.Args[2])
+	if err != nil {
+		log.Fatal("failed to get absolute path of output file")
+	}
+
+	// Read and parse go module into file object
+	mod, err := ioutil.ReadFile(goMod)
+	if err != nil {
+		log.Fatal(err)
+	}
+	file, err := modfile.Parse(goMod, mod, nil)
```

Comment:
I would feel better about using `go list` rather than parsing the go.mod file. The `go list` is the source of truth for the modules.

```
go list -m -json all | jq 'select(.Path | contains("terraform-providers"))'
```

Reply:
I am using the golang module parsing module, which is what golang uses to parse modules ;-) I am not a big fan of parsing output, since it tends to change over time and break things silently. My preference is to use the module, since it has a stable API.

Reply:
Sounds good. I was mostly concerned about doing the parsing in a bespoke way. But if we are using the same methods that are used under the covers by `go mod`, then I am satisfied.